controller/controller_test.go

```diff
@@ -1090,51 +1173,57 @@ func (er *ErrorReconciler) Reconcile(context.Context, string) error {
 }
 
 func TestStartAndShutdownWithErroringWork(t *testing.T) {
+	const testTimeout = 500 * time.Millisecond
+
+	item := types.NamespacedName{Namespace: "", Name: "bar"}
+
 	t.Cleanup(ClearAll)
-	r := &ErrorReconciler{}
-	reporter := &FakeStatsReporter{}
-	impl := NewImplWithStats(r, TestLogger(t), "Testing", reporter)
 
-	ctx, cancel := context.WithCancel(context.Background())
-	t.Cleanup(cancel)
-	doneCh := make(chan struct{})
+	impl := NewImplWithStats(&ErrorReconciler{}, TestLogger(t), "Testing", &FakeStatsReporter{})
+	impl.EnqueueKey(item)
 
-	impl.EnqueueKey(types.NamespacedName{Namespace: "", Name: "bar"})
+	ctx, cancel := context.WithTimeout(context.Background(), testTimeout)
+	t.Cleanup(cancel)
 
+	doneCh := make(chan struct{})
 	go func() {
 		defer close(doneCh)
 		// StartAll blocks until all the worker threads finish, which shouldn't
 		// be until we cancel the context.
 		StartAll(ctx, impl)
 	}()
 
-	select {
-	case <-time.After(1 * time.Second):
-		// We don't expect completion before the context is cancelled,
-		// but the workers should spin on the erroring work.
-
-	case <-doneCh:
-		t.Error("StartAll finished early.")
-	}
-
-	// By cancelling the context all the workers should complete and
-	// we should close the doneCh.
-	cancel()
+	// Keep checking the number of requeues, send to channel to indicate success.
+	itemRequeued := make(chan struct{})
+	defer close(itemRequeued)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(5 * time.Millisecond):
+				// Check that the work was requeued in RateLimiter, as NumRequeues
+				// can't fully reflect the real state of queue length.
+				// Here we need to wait for NumRequeues to be more than 1, to ensure
+				// the key get re-queued and reprocessed as expect.
+				if impl.WorkQueue().NumRequeues(item) > 1 {
+					itemRequeued <- struct{}{}
```

Comment:
If this is only expected to happen once, generally the pattern in Go is to `close()` the channel instead of sending the empty struct.  If it can happen multiple times, then this is cool too.

Reply:
That's what I would usually do as well, but there is a dangerous `defer close(itemRequeued)` which causes a panic if I close the channel twice.

Do you think I should rather leave the channel open in case of failure and let it die with the process?